### PR TITLE
Removing the previous fix of kicking off a separate thread as it's to…

### PR DESCRIPTION
**Pull request by MushMal**
…o expensive for some of the scenarios where incoming data can happen at high rate. Instead, fixing ICE agent interlocking.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/source/Ice/TurnConnection.c [hunk 1]
```diff
@@ -399,8 +403,8 @@ STATUS turnConnectionHandleChannelDataTcpMode(PTurnConnection pTurnConnection, P
 
     pCurPos = pBuffer;
     remainingBufLen = bufferLen;
-    while(remainingBufLen > 0) {
-        if (pTurnConnection->currRecvDataLen > 0) {
+    while(remainingBufLen != 0) {
```
Comment (staowyze) [inline src/source/Ice/TurnConnection.c]: probably not a big difference, may I know why changing from `>0` to `!=0`?

File: src/source/Ice/TurnConnection.c [hunk 2]
```diff
@@ -160,6 +160,10 @@ STATUS turnConnectionIncomingDataHandler(UINT64 customData, PSocketConnection pS
         CHK_STATUS(turnConnectionHandleChannelDataTcpMode(pTurnConnection, pBuffer, bufferLen));
     }
 
+    // Unlock before checking if need to deliver to data channel callback
+    MUTEX_UNLOCK(pTurnConnection->lock);
```
Comment (staowyze) [inline src/source/Ice/TurnConnection.c]: I guess this is typo, mutex is released in cleanup, a couple of lines below.

File: src/source/Ice/IceAgent.c [hunk 3]
```diff
@@ -1304,10 +1305,24 @@ STATUS incomingDataHandler(UINT64 customData, PSocketConnection pSocketConnectio
         locked = FALSE;
 
         pIceAgent->iceAgentCallbacks.inboundPacketFn(pIceAgent->customData, pBuffer, bufferLen);
-        CHK(FALSE, retStatus);
-    }
+    } else {
+        CHK_STATUS(handleStunPacket(pIceAgent, pBuffer, bufferLen, pSocketConnection, pSrc, pDest,
+                &pStunResponse, &useTurn));
+
+        if (pStunResponse != NULL) {
```
Comment (staowyze) [inline src/source/Ice/IceAgent.c]: this way is definitely better than creating threads.

File: src/source/Ice/IceAgent.c [hunk 4]
```diff
@@ -1304,10 +1305,24 @@ STATUS incomingDataHandler(UINT64 customData, PSocketConnection pSocketConnectio
...
+            locked = FALSE;
 
-    CHK_STATUS(handleStunPacket(pIceAgent, pBuffer, bufferLen, pSocketConnection, pSrc, pDest));
+            CHK_STATUS(iceUtilsSendStunPacket(pStunResponse,
+                    (PBYTE) pIceAgent->localPassword,
+                    (UINT32) STRLEN(pIceAgent->localPassword) * SIZEOF(CHAR),
+                    pSrc,
+                    pSocketConnection,
+                    pIceAgent->pTurnConnection,
+                    useTurn));
```
Comment (staowyze) [inline src/source/Ice/IceAgent.c]: dummy question, when do we need to send stun packet without using turn connection?
appears `iceUtilsSendStunPacket` is taking too many parameters.

File: src/source/Ice/TurnConnection.c (hunk 1 above)
Comment (MushMal) [reply src/source/Ice/TurnConnection.c]: Both are unsigned so won't be less than zero. It's better to be 'precise' in these cases

File: src/source/Ice/TurnConnection.c (hunk 2 above)
Comment (MushMal) [reply src/source/Ice/TurnConnection.c]: Yes, it does look weird to do it again. Let me change this. Thanks for spotting it.

File: src/source/Ice/IceAgent.c (hunk 3 above)
Comment (MushMal) [reply src/source/Ice/IceAgent.c]: Thanks. I do think that we will need a separate service thread that services all of the outgoing packets with some buffering.

File: src/source/Ice/IceAgent.c (hunk 4 above)
Comment (MushMal) [reply src/source/Ice/IceAgent.c]: This is a util function that can send stun packets via turn or direct P2P.

**Pull request merged**